### PR TITLE
users: Fix calling `last` in endless loop, fix crash on lastlog failure

### PR DESCRIPTION
**Pull request by martinpitt**
Using `[]` as "uninitialized" state for `logins` caused the useEffect() to run in an endless loop, calling `last` hundreds of times per second. It's legal for a user to not have any recent logins, in which case the effect would call `setLogins([])`, and re-triggering itself.

### CODE REVIEW

**Comment by martinpitt** [comment]
@jelly , @KKoukiou : I wonder if this is really the right way to fix it. This usage of `useEffect()` feels dangerously error-prone to me. Isn't that a use case for `useInit()`?

I updated this PR for useInit(). We have both variants in the history now, please let me know which one you prefer.

File: pkg/users/account-logs-panel.jsx [hunk 1]
```diff
@@ -18,58 +18,57 @@
...
+                    let logins = [];
+                    data.split('\n').forEach(line => {
+                        // Exclude still logged in and non user lines
+                        if (!line.includes(name) || line.includes('still')) {
+                            return;
+                        }
+                        // Exclude tmux/screen lines
+                        if (line.includes('tmux') || line.includes('screen')) {
+                            return;
+                        }
```
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:45]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-062130-a2579535-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/account-logs-panel.jsx [hunk 2]
```diff
@@ -18,58 +18,57 @@
...
-                            }
+                        // format:
+                        // admin    web console  ::ffff:172.27.0. 2021-09-24T09:02:13+00:00 - 2021-09-24T09:04:20+00:00  (00:02)
+                        const lines = line.split(/ +/);
+                        const ended = new Date(lines[lines.length - 2]);
+                        const started = new Date(lines[lines.length - 4]);
+                        const from = lines[lines.length - 5];
+                        if (isNaN(started.getTime()) || isNaN(ended.getTime())) {
+                            return;
+                        }
```
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:55]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-062130-a2579535-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/account-logs-panel.jsx [hunk 3]
```diff
@@ -18,58 +18,57 @@
...
-        }
-    }, [logins, name]);
+
+                    // Only show 15 login lines
+                    logins = logins.slice(0, 15);
+                    setLogins(logins);
+                })
+                .catch((ex) => {
+                    console.error(ex);
+                });
```
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:70]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-062130-a2579535-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
[This failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-080703-a2579535-fedora-37-pybridge/log.html#328-1) is another, unrelated crash on the login page:

> Uncaught (in promise)","lineNumber":1,"columnNumber":549430,"scriptId":"27","url":"http://127.0.0.2:9291/cockpit/@localhost/users/users.js","stackTrace":{"callFrames":[{"functionName":"wc","scriptId":"27","url":"http://127.0.0.2:9291/cockpit/@localhost/users/users.js","lineNumber":1,"columnNumber":549430}]},"exception":{"type":"object","subtype":"error",

> "className":"TypeError","description":"TypeError: Cannot read properties of undefined (reading 'split')\n    at wc (http://127.0.0.2:9291/cockpit/@localhost/users/users.js:2:549431)"

File: pkg/users/account-logs-panel.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:45]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-092548-536125bd-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/account-logs-panel.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:55]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-092548-536125bd-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/account-logs-panel.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:70]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-092548-536125bd-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/local.js [hunk 4]
```diff
@@ -101,14 +101,20 @@ function get_locked(name) {
...
-    const lastlog = await cockpit.spawn(["/usr/bin/lastlog"], { environ: ["LC_ALL=C"] })
-            .catch(err => console.warn("Unexpected error when getting last login information", err));
-    const w = await cockpit.spawn(["/usr/bin/w", "-sh"], { environ: ["LC_ALL=C"] })
-            .catch(err => console.warn("Unexpected error when getting logged in accounts", err));
+    let lastlog = [];
+    try {
+        lastlog = await cockpit.spawn(["/usr/bin/lastlog"], { environ: ["LC_ALL=C"] });
+    } catch (err) {
+        console.warn("Unexpected error when getting last login information", err);
+    }
```
Comment (cockpituous) [inline pkg/users/local.js:109]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-092548-536125bd-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/local.js [hunk 5]
```diff
@@ -101,14 +101,20 @@
...
-    const currentLogins = w.split('\n').slice(0, -1).map(line => {
-        return line.split(/ +/)[0];
-    });
+    let currentLogins = [];
+    try {
+        const w = await cockpit.spawn(["/usr/bin/w", "-sh"], { environ: ["LC_ALL=C"] });
+        currentLogins = w.split('\n').slice(0, -1).map(line => line.split(/ +/)[0]);
+    } catch (err) {
+        console.warn("Unexpected error when getting logged in accounts", err);
+    }
```
Comment (cockpituous) [inline pkg/users/local.js:117]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-092548-536125bd-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
This [unexpected message](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-092548-536125bd-fedora-37-pybridge/log.html#300) is on main already, see https://github.com/cockpit-project/cockpit/pull/18188#issuecomment-1385703320 ff. Ignoring here.

File: pkg/users/account-logs-panel.jsx [hunk 6]
```diff
@@ -18,58 +18,57 @@
...
-                        data.split('\n').forEach(line => {
-                            // Exclude still logged in and non user lines
-                            if (!line.includes(name) || line.includes('still')) {
-                                return;
-                            }
-                            // Exclude tmux/screen lines
-                            if (line.includes('tmux') || line.includes('screen')) {
-                                return;
-                            }
+    useInit(() => {
```
Comment (marusak) [inline pkg/users/account-logs-panel.jsx:33]: this does need `name` as argument. If you go to https://localhost:9090/users#/user1 and then to https://localhost:9090/users#/user2 you will still see login history from user1 as it does not recognize that it needs to re-read this

**Review by marusak — CHANGES_REQUESTED** [review]

File: pkg/users/account-logs-panel.jsx (hunk 6 above)
Comment (martinpitt) [reply pkg/users/account-logs-panel.jsx:33]: Whoops, thanks for spotting! Reproduced, fixed, validated.

File: pkg/users/account-logs-panel.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:45]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-133515-72233636-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/account-logs-panel.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:55]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-133515-72233636-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/account-logs-panel.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/users/account-logs-panel.jsx:70]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-133515-72233636-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/local.js (hunk 4 above)
Comment (cockpituous) [inline pkg/users/local.js:109]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-133515-72233636-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/users/local.js (hunk 5 above)
Comment (cockpituous) [inline pkg/users/local.js:117]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18198-20230118-133515-72233636-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by marusak — APPROVED** [review]
Thank you!

**Pull request merged**